### DISCUSSION TITLE
Is it possible to request a DCHPv6-PD without assigning it to an interface?

### DISCUSSION THREAD

**Discussion by Kentzo** [rsmarples/dhcpcd#217] [Q&A]
I'm requesting a prefix so I can distribute it among my IPsec clients (virtual IPs). This I only need dhcpcd to request and maintain binding as well as call the hook to reflect changes.

My current approach is to assign the prefix to the same interface that requested it. It works because my Router doesn't support `OPTION_PD_EXCLUDE`. But it's a unreliable long-term.

**Comment by zacknewman** [comment]
Per `dhcpcd.conf(5)`:
```bash
⋮
ia_pd [iaid [/ prefix / prefix_len] [interface [/ sla_id [/ prefix_len [/suffix]]]]]
Request a DHCPv6 Delegated Prefix for iaid.  This option must be
used in an interface block.  Unless a sla_id of 0 is assigned
with the same resultant prefix length as the delegation, a reject
route is installed for the Delegated Prefix to stop unallocated
addresses being resolved upstream.  If no interface is given then
we will assign a prefix to every other interface with a sla_id
equivalent to the interface index assigned by the OS.  Otherwise
addresses are only assigned for each interface and sla_id.  Each
assigned address will have a suffix, defaulting to 1.  If the
suffix is 0 then a SLAAC address is assigned.  You cannot assign
a prefix to the requesting interface unless the DHCPv6 server
supports the RFC 6603 Prefix Exclude Option.
⋮
```
As you can see if you don't specify any interface, then it "will assign a prefix to every other interface with a sla_id equivalent to the interface index assigned by the OS". I believe "every other interface" means the set of interfaces you have `dhcpcd` running on except the requesting interface. This means if you only have enabled the requesting interface, then you should be fine (since this set of interfaces is in fact the empty set).

The following is what my `dhcpcd.conf(5)` looks like:
```bash
router$ cat /etc/dhcpcd.conf
noipv6rs
denyinterfaces !(em0)
allowinterfaces em0
script ""
interface em0
	duid
	iaid 0
	ia_na 0
	ia_pd 0/::/60
	ipv6rs
	leasetime -1
	nooption domain_name,domain_name_servers,domain_search,fqdn,host_name,ntp_servers,time_offset,time_servers
	option rapid_commit
	persistent
	require dhcp_server_identifier
	waitip 64
```
I have *many* interfaces on my machine besides `em0`. For most of them I statically assign a /64 subnet, and I have had no issues.

**Reply by Kentzo** [reply]
My host is multi-homed (via VLANs) and dhcpcd manages most of the interfaces.

**Comment by ido** [comment]
Is there a question here?

Resolving, but feel free to reopen if you have a question.

**Reply by Kentzo** [reply]
The question is in the title. 

It doesn’t seem to be possible to request a prefix without assignment. Please consider it a feature request.

**Reply by ido** [reply]
Ah, thank you. That's what I get for working late. Reopened.

**Comment by ido** [comment]
@rsmarples I think this might be a limitation of the `dhcpcd.conf` syntax. Maybe the interfaces block of `ia_pd` should accept a `*` wildcard to specify the current default (distribute to all other interfaces) and/or a `-` (or similar) to indicate no distribution to interfaces? Seeking your advice here before I go changing things. (I'm inclined to do this in a backwards compatible way, leaving default behavior to `*`.)

@Kentzo what happens if you specify a non-existent interface name on that line? (I have no easy way to test this idea right now.)

**Reply by rsmarples** [reply]
@ido  Yeah, it's a limitation of the syntax.
We just have to document an invalid interface prefix or character.
resolvconf has such a list here: https://github.com/NetworkConfiguration/openresolv/blob/master/resolvconf.in#L806

As `-` is not allowed at the start of an interface name I suggest we just use that symbol to mean no interface assignment.

**Reply by rsmarples** [reply]
No actual code change is really needed for this. I have silenced a warning when it is used and added some documentation for it.
`-` is an invalid interface name, so it cannot exist, thus is perfectly safe to use.
